### PR TITLE
fix(galaxy-proxy): use configured Galaxy servers for version discovery

### LINKED ISSUE #261: feat(galaxy-proxy): support private Galaxy for version discovery
## Context

Galaxy version discovery in `_fetch_galaxy_versions` is hardcoded to
`https://galaxy.ansible.com`. This means the PEP 503 index always lists
versions from public Galaxy, even when the Gateway has configured private
or air-gapped Galaxy servers for collection downloads.

Flagged during review of PR #260:
https://github.com/ansible/apme/pull/260#discussion_r3049146481

Contributor acknowledged as a known limitation.

## Proposal

Derive the Galaxy API base URL from the configured Galaxy servers
(pushed from Gateway via scan metadata) rather than the hardcoded
`_GALAXY_API_URL` constant. Fall back to public Galaxy when no
configured servers are available.

### PR DESCRIPTION
**Pull request by djdanielsson**
## Summary
- Version discovery (`_fetch_galaxy_versions`) was hardcoded to `galaxy.ansible.com`, even when private Galaxy servers (console.redhat.com, Automation Hub) were configured via the Gateway UI
- Collection downloads already used configured servers, but the PEP 503 index page's version listing did not — this closes that gap

## Changes
- **`_fetch_galaxy_versions`** now accepts an optional `servers: list[GalaxyServerConfig]` parameter; tries each configured server in order (with auth tokens) before falling back to public Galaxy
- Extracted **`_fetch_versions_from`** as a single-server helper that returns `None` on failure for clean fallthrough
- **`project_page`** handler now passes the configured Galaxy servers from app state to version discovery (via `_get_galaxy_config()`)
- Added 6 tests in `TestVersionDiscoveryWithServers`: server passthrough, fallback behavior, server ordering, failure fallthrough, auth token injection, and no-token case

## Test plan
- [x] `tox -e lint` passes
- [x] `tox -e unit` passes (all 27 proxy tests pass; pre-existing failures in rule_doc_integration_test and test_session_venv_e2e are unrelated)

Closes #261

Made with [Cursor](https://cursor.com)

### CODE REVIEW

File: src/galaxy_proxy/proxy/server.py [hunk 1]
```diff
@@ -475,21 +480,75 @@ def _galaxy_version_sort_key(version: str) -> tuple[int, Version | str]:
...
+    """
+    base_urls: list[tuple[str, str | None]] = []
+    for srv in servers or []:
+        base_urls.append((srv.url.rstrip("/"), srv.token))
+    base_urls.append((_GALAXY_API_URL, None))
+
+    for base_url, token in base_urls:
+        versions = await _fetch_versions_from(namespace, name, base_url, token=token)
+        if versions is not None:
+            return sorted(set(versions), key=_galaxy_version_sort_key)
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/galaxy_proxy/proxy/server.py:516]: Good catch. Configured `GalaxyServerConfig.url` values from ansible.cfg commonly include `/api/`, `/api/galaxy/`, or `/api/automation-hub` suffixes. Concatenating those with `_GALAXY_VERSIONS_PATH` (`/api/v3/...`) produced `/api/api/v3/...` which 404s.

Added `_normalize_galaxy_url()` which strips everything from the first `/api` segment onward before appending the versions path. Parametrized test covers all common URL shapes (bare, `/api/`, `/api/galaxy/`, `/api/galaxy/content/published/`, `/api/automation-hub`). Also added a test that exercises `_fetch_versions_from` with an `/api/` URL and asserts the constructed URL doesn't contain `/api/api/`.

Fixed in dbcb268.

File: tests/test_galaxy_proxy_server.py [hunk 2]
```diff
@@ -357,6 +358,195 @@ def test_push_rejects_duplicate_name(self, app: TestClient) -> None:
...
+            assert resp.status_code == 200
+            mock_fetch.assert_called_once()
+            call_kwargs = mock_fetch.call_args
+            servers = call_kwargs.kwargs.get("servers") or call_kwargs[1].get("servers")
+            assert servers is not None
+            assert len(servers) == 1
+            assert isinstance(servers[0], GalaxyServerConfig)
+            assert servers[0].url == "https://hub.example.com"
+            assert servers[0].token == "tok"
+
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply tests/test_galaxy_proxy_server.py:394]: Added. Four new tests cover this:

1. `test_normalize_galaxy_url` — parametrized test with 6 URL shapes including `/api/`, `/api/galaxy/`, and `/api/automation-hub` variants
2. `test_fetch_versions_from_normalizes_api_url` — exercises the actual HTTP client path with `https://hub.example.com/api/` and asserts the constructed URL doesn't contain `/api/api/`
3. `test_version_discovery_with_api_url_through_project_page` — integration-style test that pushes a server config with `/api/` URL and verifies it flows through the project page handler

All 35 proxy tests pass. Fixed in dbcb268.

File: src/galaxy_proxy/proxy/server.py [hunk 3]
```diff
@@ -500,8 +580,15 @@ async def _fetch_galaxy_versions(namespace: str, name: str) -> list[str]:
...
     except (httpx.HTTPError, KeyError, ValueError) as exc:
-        logger.warning("Failed to fetch Galaxy versions for %s.%s: %s", namespace, name, exc)
-    return sorted(set(versions), key=_galaxy_version_sort_key)
+        logger.warning(
+            "Failed to fetch Galaxy versions for %s.%s from %s: %s",
+            namespace,
+            name,
+            base_url,
+            exc,
+        )
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/galaxy_proxy/proxy/server.py:603]: Good point. Demoted the per-server failure log in `_fetch_versions_from` from WARNING to DEBUG — intermediate failures are expected when trying multiple servers in order. Added a single WARNING in `_fetch_galaxy_versions` that fires only when *all* servers (including public Galaxy fallback) fail, listing which servers were tried.

Fixed in 3938e65.

File: src/galaxy_proxy/proxy/server.py [hunk 4]
```diff
@@ -475,21 +480,103 @@ def _galaxy_version_sort_key(version: str) -> tuple[int, Version | str]:
...
+    Args:
+        raw_url: Server URL as provided by the user / Gateway config.
+
+    Returns:
+        Base URL with trailing ``/api...`` segments removed.
+    """
+    url = raw_url.rstrip("/")
+    if "/api" in url:
+        url = url[: url.index("/api")]
+    return url
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/galaxy_proxy/proxy/server.py]: Good catch — `https://api.example.com` would match `/api` in the hostname and produce `https:/`. Switched to `urllib.parse.urlsplit`/`urlunsplit` so only the *path* component is inspected. Added parametrized test cases for `https://api.example.com`, `https://api.example.com/`, and `https://api.galaxy.example.com/api/` to cover this edge case.

Fixed in 9c4ff88.
